spv/sync.go

```diff
@@ -0,0 +1,1420 @@
+	FetchMissingCFiltersStarted  func(walletID int)
+	FetchMissingCFiltersProgress func(walletID int, startCFiltersHeight, endCFiltersHeight int32)
+	FetchMissingCFiltersFinished func(walletID int)
+	FetchHeadersStarted          func(peerInitialHeight int32)
+	FetchHeadersProgress         func(lastHeaderHeight int32, lastHeaderTime int64)
+	FetchHeadersFinished         func()
+	DiscoverAddressesStarted     func(walletID int)
+	DiscoverAddressesFinished    func(walletID int)
+	RescanStarted                func(walletID int)
+	RescanProgress               func(walletID int, rescannedThrough int32)
+	RescanFinished               func(walletID int)
+
+	// MempoolTxs is called whenever new relevant unmined transactions are
+	// observed and saved.
+	MempoolTxs func(walletID int, txs []*wire.MsgTx)
+
+	// TipChanged is called when the main chain tip block changes.
+	// When reorgDepth is zero, the new block is a direct child of the previous tip.
+	// If non-zero, one or more blocks described by the parameter were removed from
+	// the previous main chain.
+	// txs contains all relevant transactions mined in each attached block in
+	// unspecified order.
+	// reorgDepth is guaranteed to be non-negative.
+	TipChanged func(tip *wire.BlockHeader, reorgDepth int32, txs []*wire.MsgTx)
+}
+
+// NewSyncer creates a Syncer that will sync the wallet using SPV.
+func NewSyncer(wallets map[int]*wallet.Wallet, chainParams *chaincfg.Params, lp *p2p.LocalPeer, discoverAccountsIfUnlocked bool) *Syncer {
+	rescanFilter := make(map[int]*wallet.RescanFilter)
+	filterData := make(map[int]*blockcf.Entries)
+	for walletID := range wallets {
+		rescanFilter[walletID] = wallet.NewRescanFilter(nil, nil)
+		filterData[walletID] = &blockcf.Entries{}
+	}
+
+	return &Syncer{
+		chainParams:                chainParams,
+		wallets:                    wallets,
+		discoverAccountsIfUnlocked: discoverAccountsIfUnlocked,
+		loadedFilters:              make(map[int]bool, len(wallets)),
+		connectingRemotes:          make(map[string]struct{}),
+		remotes:                    make(map[string]*p2p.RemotePeer),
+		rescanFilter:               rescanFilter,
+		filterData:                 filterData,
+		filterWalletID:             -1,
+		seenTxs:                    lru.NewCache(2000),
+		lp:                         lp,
+	}
+}
+
+// SetPersistentPeers sets each peer as a persistent peer and disables DNS
+// seeding and peer discovery.
+func (s *Syncer) SetPersistentPeers(peers []string) {
+	s.persistentPeers = peers
+}
+
+// SetNotifications sets the possible various callbacks that are used
+// to notify interested parties to the syncing progress.
+func (s *Syncer) SetNotifications(ntfns *Notifications) {
+	s.notifications = ntfns
+}
+
+// synced checks the atomic that controls wallet syncness and if previously
+// unsynced, updates to synced and notifies the callback, if set.
+func (s *Syncer) synced(walletID int) {
+	if s.notifications != nil && s.notifications.Synced != nil {
```

Comment:
For this `synced` method and the next `unsynced` method, the atomic swap checks were removed. If instead of a single `s.atomicWalletSynced` property, we have a map of `atomicWalletSynced` properties for each wallet, we can retain this functionality.

Instead of `atomicWalletSynced   uint32` in the `Syncer` struct, we can have `atomicWalletsSynced   map[int]*uint32`. Using pointers ensure that when changes from atomic swaps are retained in the map.

Then comparison and swap can be done thus:
`if atomic.CompareAndSwapUint32(s.atomicWalletsSynced[walletID], 0, 1)`